src/sentry/auth/staff.py

```diff
@@ -0,0 +1,319 @@
+            )
+            return
+
+        if not cookie_token:
+            if data:
+                logger.warning(
+                    "staff.missing-cookie-token",
+                    extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+                )
+            return
+        elif not data:
+            logger.warning(
+                "staff.missing-session-data",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        session_token = data.get("tok")
+        if not session_token:
+            logger.warning(
+                "staff.missing-session-token",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        if not constant_time_compare(cookie_token, session_token):
+            logger.warning(
+                "staff.invalid-token",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        if data["uid"] != str(request.user.id):
+            logger.warning(
+                "staff.invalid-uid",
+                extra={
+                    "ip_address": request.META["REMOTE_ADDR"],
+                    "user_id": request.user.id,
+                    "expected_user_id": data["uid"],
+                },
+            )
+            return
+
+        if current_datetime is None:
+            current_datetime = timezone.now()
+
+        try:
+            data["idl"] = datetime.utcfromtimestamp(float(data["idl"])).replace(tzinfo=timezone.utc)
+        except (TypeError, ValueError):
+            logger.warning(
+                "staff.invalid-idle-expiration",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+                exc_info=True,
+            )
+            return
+
+        if data["idl"] < current_datetime:
+            logger.info(
+                "staff.session-expired",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        try:
+            data["exp"] = datetime.utcfromtimestamp(float(data["exp"])).replace(tzinfo=timezone.utc)
+        except (TypeError, ValueError):
+            logger.warning(
+                "staff.invalid-expiration",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+                exc_info=True,
+            )
+            return
+
+        if data["exp"] < current_datetime:
+            logger.info(
+                "staff.session-expired",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        return data
+
+    def _populate(self, current_datetime=None):
```

Comment:
Could we name this function to something else? Populate is very generic, I'm not sure what it's supposed to be populating, and if it's okay to rerun this function or add to another private function. 

If this is only meant to be run once, could we add some guard rails/checks for that? Or restructure this function so that it returns the values we want to populate and the `__init__` can be responsible for adding them to the class?